src/main/java/org/terasology/machines/systems/MachineCommonSystem.java

```diff
@@ -49,9 +50,12 @@ public void onNestedMachineDefinitionAdded(OnAddedComponent event, EntityRef ent
     }
 
     private void addProcessingMachine(EntityRef entity, MachineDefinitionComponent machineDefinition) {
+        RetainComponentsComponent retainComponents = new RetainComponentsComponent();
 
         // configure the input/output inventories
         if (!entity.hasComponent(InventoryComponent.class) && machineDefinition.inputSlots + machineDefinition.requirementSlots + machineDefinition.outputSlots > 0) {
+            retainComponents.components.add(InventoryComponent.class);
+            entity.addOrSaveComponent(retainComponents);
```

Comment:
I think it should suffice to only `addOrSave` the component once at the end. You can build up the final state by adding elements to the set and just save it once.

Reply:
done